### PR TITLE
Add support for lifetime dependence in parameter position

### PR DESCRIPTION
**Pull request by meg-gupta**
Extend support for lifetime dependence in parameter position:

```swift
func mayReassign(span: dependsOn(a) inout Span<Int>, to a: ContiguousArray<Int>) {
  span = a.span()
}
```
https://github.com/swiftlang/swift-evolution/blob/e814e62c17741dfe82e2f55eb1d751157c77a853/proposals/NNNN-lifetime-dependency.md#dependent-parameters contains a detailed explanation

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Review by atrick — APPROVED** [review]
Wow! Lots of great work here that we need to unblock the proposal and downstream development. This PR does not need to be perfect. I want this to land today so it has time to go through a CI cycle while you are available.

Is there a rebase problem with the "Update Array bounds check optimization" commit?

Note on mangling:

We probably want a different mangling, but we can follow up with that in a separate PR (I can file a bug if you agree). We will want to support component lifetimes. So each dependence will be a parameter index, potentially followed by a list of component names. e.g.

    func foo(arg0, arg1) -> dependsOn(arg1.a.b) R

Would be mangled something like:

    Yli0_1a1b

(maybe with some additional character to signal presence or absence of components).

We do not need to use INDEX-SUBSET because a dependence will really almost never be more than one index. We do want to be able to mangle multiple dependence sources on the same target, even though it's extremely unlikely. But it seems that we should just repeat the mangling prefix in that case:

    func foo(arg0, arg1) -> dependsOn(arg0, arg1.a.b) R

Would be mangled something like:

    Yli_Yli0_1a1b

@eeckstein  may want to look at the mangling commit and give suggestions.

Again, please don't hold up this PR for any mangling improvements.

File: SwiftCompilerSources/Sources/SIL/FunctionConvention.swift
Comment (eeckstein) [inline SwiftCompilerSources/Sources/SIL/FunctionConvention.swift:84]: You enumerate but don't use the index... looks like something's wrong here

File: SwiftCompilerSources/Sources/SIL/FunctionConvention.swift
Comment (eeckstein) [inline SwiftCompilerSources/Sources/SIL/FunctionConvention.swift:248]: SIL APIs shouldn't expose bridged data types. Please either define a swift struct around `BridgedLifetimeDependenceInfo` or - if you only use `LifetimeDependencies` internally in function conventions, don't make it public.

File: lib/AST/ASTMangler.cpp
Comment (eeckstein) [inline lib/AST/ASTMangler.cpp:3229]: Do we need some deployment target checks here so that we don't run into the same back-deployment runtime crash as with isolated function types (https://github.com/swiftlang/swift/pull/74916)?
For example, if a struct contains a closure with the new lifetime dependencies, then an old runtime cannot demangle such a type.

Comment (meg-gupta) [reply lib/AST/ASTMangler.cpp:3229]: Looks like we definitely need that. Thanks for the pointer.

File: test/SIL/explicit_lifetime_dependence_specifiers.swift
Comment (meg-gupta) [reply test/SIL/explicit_lifetime_dependence_specifiers.swift:26]: This CHECK line should actually move up to line 11. And we need a new CHECK line here.

File: lib/AST/LifetimeDependence.cpp
Comment (xedin) [inline lib/AST/LifetimeDependence.cpp:580]: Same as below about `std::for_each`

**Review by xedin — APPROVED** [review]
LGTM!

File: lib/AST/LifetimeDependence.cpp
Comment (meg-gupta) [reply lib/AST/LifetimeDependence.cpp]: Done

File: lib/AST/LifetimeDependence.cpp
Comment (meg-gupta) [reply lib/AST/LifetimeDependence.cpp]: Done

File: SwiftCompilerSources/Sources/SIL/FunctionConvention.swift
Comment (meg-gupta) [reply SwiftCompilerSources/Sources/SIL/FunctionConvention.swift]: Done

File: SwiftCompilerSources/Sources/SIL/FunctionConvention.swift
Comment (meg-gupta) [reply SwiftCompilerSources/Sources/SIL/FunctionConvention.swift]: Fixed

File: lib/AST/ASTMangler.cpp
Comment (meg-gupta) [reply lib/AST/ASTMangler.cpp:3229]: Included this in the new change. No metadata test because we don't allow lifetime dependencies on function types yet.

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@atrick @eeckstein I'll make the suggested mangling changes as a follow on.

**Review by eeckstein — APPROVED** [review]

**Comment by DougGregor** [comment]
> rdar://129862031

Please provide a comment that describes what this pull request does, because radar numbers aren't useful outside of Apple.

File: lib/Serialization/Deserialization.cpp
Comment (DougGregor) [inline lib/Serialization/Deserialization.cpp:7700]: I think you need a `ctx.AllocateCopy(lifetimeDependencies)` here?

File: lib/Serialization/Deserialization.cpp
Comment (DougGregor) [inline lib/Serialization/Deserialization.cpp:7073]: I think you need to `ctx.AllocateCopy(lifetimeDependencies)` here

**Comment by DougGregor** [comment]
I do have a mangling question... I assume that we're not going to allow overloading based on the lifetime dependencies of a function. In that case, do we need to have a mangling for all of the lifetime dependencies?

**Comment by meg-gupta** [comment]
@DougGregor Yeah, we won't be overloading on lifetime dependencies. I think we do need to mangle all lifetime dependencies (result, inout parameters, self at least) because adding or changing them is abi breaking because we need the correct `mark_dependence` SIL instructions in the caller. So, mangling them will ensure we don't end up with the same symbol when lifetime dependencies change.

File: lib/Serialization/Deserialization.cpp
Comment (meg-gupta) [reply lib/Serialization/Deserialization.cpp:7073]: Yes, thank you!

**Comment by DougGregor** [comment]
> @DougGregor Yeah, we won't be overloading on lifetime dependencies. I think we do need to mangle all lifetime dependencies (result, inout parameters, self at least) because adding or changing them is abi breaking since we need the correct `mark_dependence` SIL instructions in the caller. So, mangling them will ensure we don't end up with the same symbol when lifetime dependencies change.

I thought it would be ABI-compatible to remove lifetime dependencies, and that might even be a likely course of evolution: an API might initially not document its dependencies, and rely on the the inferred dependencies. Then later they realize that the inferred dependencies were too broad and document the narrower lifetime dependencies. We can cope with this via `@_silgen_name` or some bespoke attribute to disable this part of the mangling, but I think it's worth considering whether it's more effort to mangle this information vs. dropping it. Not all aspects of ABI are described in the mangling.

**Comment by meg-gupta** [comment]
>I thought it would be ABI-compatible to remove lifetime dependencies, and that might even be a likely course of evolution: an API might initially not document its dependencies, and rely on the the inferred dependencies. Then later they realize that the inferred dependencies were too broad and document the narrower lifetime dependencies. We can cope with this via @_silgen_name or some bespoke attribute to disable this part of the mangling, but I think it's worth considering whether it's more effort to mangle this information vs. dropping it. Not all aspects of ABI are described in the mangling.

Okay - I thought we want to mangle all ABI breaking changes.
Yes, removing dependencies is not abi breaking.
I'll take a look at what `@_silgen_name` and explore what that involves before doing the follow up on mangling. Thanks for the suggestion.

**Comment by atrick** [comment]
> I thought it would be ABI-compatible to remove lifetime dependencies, and that might even be a likely course of evolution: an API might initially not document its dependencies, and rely on the the inferred dependencies. Then later they realize that the inferred dependencies were too broad and document the narrower lifetime dependencies. We can cope with this via `@_silgen_name` or some bespoke attribute to disable this part of the mangling, but I think it's worth considering whether it's more effort to mangle this information vs. dropping it. Not all aspects of ABI are described in the mangling.

When a function result is nonescapable, we always mangle at least one dependence. There could be more than one dependence, and the source of that dependence could generally be any argument, so the ABI is sensitive to those changes.

Generally, we want any aspect of the ABI that is sensitive to source changes to be mangled simply to catch problems at link time rather than debugging miscompiles.

If the client adds `~Escapable` to an argument type, then we want to catch any possibility that the API did not know about that constraint when it was compiled.

If the client removes `~Escapable` from the return type, then we want to catch any possibility that the API assumed that constraint when it was compiled.

**Comment by atrick** [comment]
@meg-gupta, just to be sure we're on the same page...

For nonescapable results, lifetime dependence inference rules are independent of mangling. We always mangle whether it could be inferred or not. We could change/fix inferrence rules without risk of miscompile.

`inout` arguments are a little different. In the callee, we always assume a dependence on the incoming value. The outgoing value is always valid if it only depends on its incoming value. We allow the API to specify that an `inout` argument has additional dependencies on other arguments. Those additional dependencies need to be represented and mangled. But we have no way of suppressing the inout argument's dependence on itself.

File: lib/Serialization/Deserialization.cpp
Comment (meg-gupta) [reply lib/Serialization/Deserialization.cpp]: Fixed

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
>@meg-gupta, just to be sure we're on the same page...

For nonescapable results, lifetime dependence inference rules are independent of mangling. We always mangle whether it could be inferred or not. We could change/fix inferrence rules without risk of miscompile.

Yes, we should mangle/serialize/ast print both inferred and explicit dependencies. No confusion here.

**Comment by atrick** [comment]
`ModuleInterface/swift_build_sdk_interfaces/find-modules.test-sh` appears to be flaky on Windows ☹️

**Comment by atrick** [comment]
@swift-ci test Windows Platform

**Pull request merged**
